<font color='blue'>Привет! Поздравляю с первым выполненным проектом. Проделана большая работа. Я проверю твою работу и напишу свои комментарии. Они, в зависимости от результата проверки, будут в </font><font color='green'>зеленом (когда все отлично),</font> <font color='brown'>коричневом (когда можно лучше)</font><font color='blue'> или</font> <font color='red'>красном (когда нужно переделать, чтобы я принял работу).</font><font color='blue'> Ну и в синем будут просто советы и нейтральные комментарии. Постарайся учесть все это в дальнейших проектах.</font>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/data.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Изучим каждый столбец по-отдельности

In [2]:
print(data.groupby('children').count())
print(data.groupby('days_employed').count())

          days_employed  dob_years  education  education_id  family_status  \
children                                                                     
-1                   44         47         47            47             47   
 0                12710      14149      14149         14149          14149   
 1                 4343       4818       4818          4818           4818   
 2                 1851       2055       2055          2055           2055   
 3                  294        330        330           330            330   
 4                   34         41         41            41             41   
 5                    8          9          9             9              9   
 20                  67         76         76            76             76   

          family_status_id  gender  income_type   debt  total_income  purpose  
children                                                                       
-1                      47      47           47     47     

In [3]:
print(data[data['days_employed'] < 0].count())

children            15906
days_employed       15906
dob_years           15906
education           15906
education_id        15906
family_status       15906
family_status_id    15906
gender              15906
income_type         15906
debt                15906
total_income        15906
purpose             15906
dtype: int64


In [4]:
print(data.groupby('dob_years').count())

           children  days_employed  education  education_id  family_status  \
dob_years                                                                    
0               101             91        101           101            101   
19               14             13         14            14             14   
20               51             46         51            51             51   
21              111             93        111           111            111   
22              183            166        183           183            183   
23              254            218        254           254            254   
24              264            243        264           264            264   
25              357            334        357           357            357   
26              408            373        408           408            408   
27              493            457        493           493            493   
28              503            446        503           503     

In [5]:
print(data.groupby('education').count())

                     children  days_employed  dob_years  education_id  \
education                                                               
ВЫСШЕЕ                    274            251        274           274   
Высшее                    268            243        268           268   
НАЧАЛЬНОЕ                  17             16         17            17   
НЕОКОНЧЕННОЕ ВЫСШЕЕ        29             22         29            29   
Начальное                  15             14         15            15   
Неоконченное высшее        47             40         47            47   
СРЕДНЕЕ                   772            705        772           772   
Среднее                   711            646        711           711   
УЧЕНАЯ СТЕПЕНЬ              1              1          1             1   
Ученая степень              1              1          1             1   
высшее                   4718           4222       4718          4718   
начальное                 250            231       

In [6]:
print(data.groupby('family_status').count())

                       children  days_employed  dob_years  education  \
family_status                                                          
Не женат / не замужем      2813           2525       2813       2813   
в разводе                  1195           1083       1195       1195   
вдовец / вдова              960            865        960        960   
гражданский брак           4177           3735       4177       4177   
женат / замужем           12380          11143      12380      12380   

                       education_id  family_status_id  gender  income_type  \
family_status                                                                
Не женат / не замужем          2813              2813    2813         2813   
в разводе                      1195              1195    1195         1195   
вдовец / вдова                  960               960     960          960   
гражданский брак               4177              4177    4177         4177   
женат / замужем            

In [7]:
print(data.groupby('gender').count())

        children  days_employed  dob_years  education  education_id  \
gender                                                                
F          14236          12752      14236      14236         14236   
M           7288           6598       7288       7288          7288   
XNA            1              1          1          1             1   

        family_status  family_status_id  income_type   debt  total_income  \
gender                                                                      
F               14236             14236        14236  14236         12752   
M                7288              7288         7288   7288          6598   
XNA                 1                 1            1      1             1   

        purpose  
gender           
F         14236  
M          7288  
XNA           1  


In [8]:
print(data.groupby('income_type').count())

                 children  days_employed  dob_years  education  education_id  \
income_type                                                                    
безработный             2              2          2          2             2   
в декрете               1              1          1          1             1   
госслужащий          1459           1312       1459       1459          1459   
компаньон            5085           4577       5085       5085          5085   
пенсионер            3856           3443       3856       3856          3856   
предприниматель         2              1          2          2             2   
сотрудник           11119          10014      11119      11119         11119   
студент                 1              1          1          1             1   

                 family_status  family_status_id  gender   debt  total_income  \
income_type                                                                     
безработный                  2       

In [9]:
print(data.groupby('debt').count())

      children  days_employed  dob_years  education  education_id  \
debt                                                                
0        19784          17780      19784      19784         19784   
1         1741           1571       1741       1741          1741   

      family_status  family_status_id  gender  income_type  total_income  \
debt                                                                       
0             19784             19784   19784        19784         17780   
1              1741              1741    1741         1741          1571   

      purpose  
debt           
0       19784  
1        1741  


In [10]:
print(data.groupby('total_income').count())

              children  days_employed  dob_years  education  education_id  \
total_income                                                                
2.066726e+04         1              1          1          1             1   
2.120528e+04         1              1          1          1             1   
2.136765e+04         1              1          1          1             1   
2.169510e+04         1              1          1          1             1   
2.189561e+04         1              1          1          1             1   
...                ...            ...        ...        ...           ...   
1.711309e+06         1              1          1          1             1   
1.715018e+06         1              1          1          1             1   
1.726276e+06         1              1          1          1             1   
2.200852e+06         1              1          1          1             1   
2.265604e+06         1              1          1          1             1   

In [11]:
print(data.groupby('purpose').count())

                                        children  days_employed  dob_years  \
purpose                                                                      
автомобили                                   478            421        478   
автомобиль                                   495            454        495   
высшее образование                           453            413        453   
дополнительное образование                   462            414        462   
жилье                                        647            587        647   
заняться высшим образованием                 496            440        496   
заняться образованием                        412            357        412   
на покупку автомобиля                        472            442        472   
на покупку подержанного автомобиля           479            437        479   
на покупку своего автомобиля                 505            452        505   
на проведение свадьбы                        777            685 

In [12]:
print(data[data['dob_years'] == 62])

       children  days_employed  dob_years education  education_id  \
30            1  335581.668515         62   среднее             1   
71            0  338113.529892         62   среднее             1   
105           0   -2098.626296         62   СРЕДНЕЕ             1   
145           0            NaN         62   среднее             1   
162           0   -3755.843042         62   среднее             1   
...         ...            ...        ...       ...           ...   
21057         0            NaN         62   среднее             1   
21108         0  349269.730785         62   среднее             1   
21150         0  330569.551611         62   Среднее             1   
21240         0  344989.762566         62   среднее             1   
21508         0  386497.714078         62   среднее             1   

          family_status  family_status_id gender income_type  debt  \
30      женат / замужем                 0      F   пенсионер     0   
71      женат / замужем        

### Вывод

1. Таблица состоит из 12 столбцов и 21525 строк. 
2. В таблице представлены данные по 21525 клиенту банка. 
3. В столбцах 'days_employed' и 'total_income' имеются пропуски.
4. Количество детей: -1, 0, 1, 2, 3, 4, 5, 20. Количество -1 и 20 ошибки.
5. Столбец 'days_employed' заполнен некорректно, так как даны не целые цисла и 15906 имеют отрицательное значение.
6. Возраст клиентов 0, 19-75. Возраст "0" является ошибкой.
7. Образование клиентов - начальное, среднее, неоконченное высшее, высшее, ученая степень.
8. Типы занятости - безработный, в декрете, госслужащий, компаньон, пенсионер, предприниматель, сотрудник, студент.


### Шаг 2. Предобработка данных

### Обработка пропусков

In [82]:
data_na = (data[data['total_income'].isnull()])
print(data_na.groupby('income_type')['income_type'].count())
print(data.groupby('income_type')['income_type'].count())
def mediana_income_type(type):
    med = data.loc[data['income_type'] == type, 'total_income'].median()
    data.loc[data['income_type'] == type, 'total_income'] = data.loc[data['income_type'] == type, 'total_income'].fillna(med)
    return data
income_type = []
for row in data['income_type'].unique():
    income_type.append(row)
for row in income_type:
    mediana_income_type(row)
data.info()

income_type
госслужащий         147
компаньон           508
пенсионер           413
предприниматель       1
сотрудник          1105
Name: income_type, dtype: int64
income_type
безработный            2
в декрете              1
госслужащий         1459
компаньон           5085
пенсионер           3856
предприниматель        2
сотрудник          11119
студент                1
Name: income_type, dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             2152

### Вывод

1. Из анализа пропусков следует, что пропуски в столбце 'days_employed' совпадают с пропусками в столбце 'total_income'. Количество пропусков - 2174.
2. Гипотеза о том, что пропуски в доходах относятся только к одной группе занятых: например пенсионерам, не подтвердилась. Пропуски присутствуют почти во всех типах занятости.
3. Гипотеза о том, что некоторых типах занятости в принципе не указываются доходы: например госслужащие, также не подтвердилась. Пропуски составляют около 10 % от общего количества типа занятости. 
4. Скорее всего данные отсутствуют из-за не предоставления клиентом данных по исследуемым характеритсикам, либо отсутствие данных по страховым взносам в ПФР, либо непредоставление справки 2 НДФЛ.
5. Для того, чтобы сохранить остальные характеристики в данных строках, пропуски заменены на медианы доходов по роду деятельности.

### Замена типа данных

In [83]:
data['total_income'] = data['total_income'].astype(int)
data = data.drop(data[data['children'] == -1].index).reset_index(drop = True)
data = data.drop(data[data['children'] == 20].index).reset_index(drop = True)
data = data.drop(data[data['dob_years'] == 0].index).reset_index(drop = True)
data = data.drop(data[data['gender'] == 'XNA'].index).reset_index(drop = True)
children = data['children']
years = data['dob_years']
try:
    for row in children:
        x = row / ((row + 1) + abs(row + 1))
except:
    print('проверь количество детей')
try:
    for row in years:
        y = row / row
except:
    print('проверь возраст клиента')


### Вывод

1. Тип данных столбца 'total_income' переведен из вещественного в целочисленный методом astype().
2. Так как доля клиентов с количеством детей "-1" и "20" составляет менее 1% от общего количества, данные строки были удалены.
3. Так как доля клиентов с возрастом "0" составляет менее 1% от общего количества, данные строки были удалены.
4. Клиент без указания пола был удален.
5. Применен метеод try-except для предотвращения ошибок при указании количества детей и возраста клиента.

### Обработка дубликатов

In [84]:
data['education'] = data['education'].str.lower()
#print(data[data.duplicated(keep=False)].count())
data = data.drop_duplicates().reset_index(drop = True)
print(data[data.duplicated(keep=False)].count())


children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


### Вывод

1. Для нахождения дубликатов строки в столбце 'education' приведены к нижнему регистру.
2. Дубликаты удалены методом drop_duplicates() и проведена переиндексация строк.
3. Причинами появления дубликатов могут быть:
- ошибка оператора, заполняющего таблицу
- клиент обращался в несколько отделений банка
- ошибка сотрудника, собирающего таблицу.

### Лемматизация

In [85]:
from pymystem3 import Mystem
m = Mystem()
purpose_unique = data['purpose'].unique()
lemmas = []
for row in purpose_unique:
    lemmas += m.lemmatize(row)
from collections import Counter
print(Counter(lemmas))
category_purpose = ['свадьба', 'недвижимость', 'автомобиль', 'образование']
def grouping_purpose(purpose):
    lemmas = m.lemmatize(purpose)
    for lemma in lemmas:
        if lemma == 'жилье':
            return 'недвижимость'
        else:
            if lemma in category_purpose:
                return lemma
data['category_purpose'] = data['purpose'].apply(grouping_purpose)
data


Counter({' ': 59, '\n': 38, 'покупка': 10, 'недвижимость': 10, 'автомобиль': 9, 'образование': 9, 'жилье': 7, 'с': 5, 'операция': 4, 'на': 4, 'свой': 4, 'свадьба': 3, 'строительство': 3, 'получение': 3, 'высокий': 3, 'дополнительный': 2, 'для': 2, 'коммерческий': 2, 'жилой': 2, 'заниматься': 2, 'сделка': 2, 'приобретение': 1, 'сыграть': 1, 'проведение': 1, 'семья': 1, 'собственный': 1, 'подержать': 1, 'со': 1, 'подержанный': 1, 'профильный': 1, 'сдача': 1, 'ремонт': 1})


children  days_employed  dob_years education  education_id  \
0             1   -8437.673028         42    высшее             0   
1             1   -4024.803754         36   среднее             1   
2             0   -5623.422610         33   среднее             1   
3             3   -4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21225         1   -4529.316663         43   среднее             1   
21226         0  343937.404131         67   среднее             1   
21227         1   -2113.346888         38   среднее             1   
21228         3   -3112.481705         38   среднее             1   
21229         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21225  гражданский брак                 1      F   компаньон     0   
21226   женат / замужем                 0      F   пенсионер     0   
21227  гражданский брак                 1      M   сотрудник     1   
21228   женат / замужем                 0      M   сотрудник     1   
21229   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose category_purpose  
0            253875                 покупка жилья     недвижимость  
1            112080       приобретение автомобиля       автомобиль  
2            145885                 покупка жилья     недвижимость  
3            267628    дополнительное образование      образование  
4            158616               сыграть свадьбу          свадьба  
...             ...                           ...              ...  
21225        224791             операции с жильем     недвижимость  
21226        155999          сделка с автомобилем       автомобиль  
21227         89672                  недвижимость     недвижимость  
21228        244093  на покупку своего автомобиля       автомобиль  
21229         82047         на покупку автомобиля       автомобиль  

[21230 rows x 13 columns]

### Вывод

1. Выполнен процесс лемматизации:
 - сохранены уникальные цели получения кредита
 - проведена лемматизация для данных целей
 - леммы, на которые можно разбить цели: свадьба, недвижимость, автомобиль, образование.
2. Добавлен дополнительный столбец с категорией цели.

### Категоризация данных

In [86]:
def grouping_children(amount):
    if amount == 0:
        return 'нет детей'
    else:
        return 'есть дети'
data['grouping_children'] = data['children'].apply(grouping_children)
def grouping_family(status):
    if status == 'гражданский брак' or status == 'женат / замужем':
        return 'семейный'
    else:
        return 'одинокий'
data['grouping_family'] = data['family_status'].apply(grouping_family)
def grouping_total_income(income):
    points = []
    quantiles = data['total_income'].quantile([0.2, 0.4, 0.6, 0.8]).astype(int)
    for row in quantiles:
        points.append(row)
    if income <= points[0]:
        return 'низкий доход'
    elif points[0] < income <= points[1]:
        return 'доход ниже среднего'
    elif points[1] < income <= points[2]:
        return 'средний доход'
    elif points[2] < income <= points[3]:
        return 'доход выше среднего'
    else:
        return 'высокий доход'
data['grouping_total_income'] = data['total_income'].apply(grouping_total_income)
data_grouping = data[['debt', 'category_purpose', 'grouping_children', 'grouping_family', 'grouping_total_income']]
data_grouping


debt category_purpose grouping_children grouping_family  \
0         0     недвижимость         есть дети        семейный   
1         0       автомобиль         есть дети        семейный   
2         0     недвижимость         нет детей        семейный   
3         0      образование         есть дети        семейный   
4         0          свадьба         нет детей        семейный   
...     ...              ...               ...             ...   
21225     0     недвижимость         есть дети        семейный   
21226     0       автомобиль         нет детей        семейный   
21227     1     недвижимость         есть дети        семейный   
21228     1       автомобиль         есть дети        семейный   
21229     0       автомобиль         есть дети        семейный   

      grouping_total_income  
0             высокий доход  
1       доход ниже среднего  
2             средний доход  
3             высокий доход  
4             средний доход  
...                     ...  
21225         высокий доход  
21226         средний доход  
21227          низкий доход  
21228         высокий доход  
21229          низкий доход  

[21230 rows x 5 columns]

### Вывод

1. Цели кредита сгруппированы в предыдущем блоке на следующие типы: 'свадьба', 'недвижимость', 'автомобиль', 'образование'.
2. По наличию детей клиенты объединены в группы 'есть дети' и 'нет детей'.
3. По семейному положению клиенты разделены на 'семейных' и 'одиноких'. К семейным отнесены клиенты замужем/женат и в гражданском браке. Остальные клиенты отнесены к одиноким.
4. По ежемесячному доходу доходы клиентов поделены на низкий доход, доход ниже среднего, средний доход, доход выше среднего, высокий доход. Разделение выполнено с помощью DataFrame.quantile. 
5. Данные группировки добавлены в общую таблицу в виде дополинтельных столбцов.
6. Для удобства новые категории клиентов и информация о задолженности вынесены в отдельную таблицу data_grouping.
7. По ежемесячному доходу доходы клиентов поделены на низкий доход, доход ниже среднего, средний доход, доход выше среднего, высокий доход. Разделение выполнено с помощью DataFrame.quantile.

### Шаг 3. Ответы на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [87]:
debt = data.groupby('children')['debt'].sum()
total = data.groupby('children')['children'].count()
print(debt / total)


children
0    0.075458
1    0.092028
2    0.095145
3    0.082317
4    0.097561
5    0.000000
dtype: float64


### Вывод

Проанализировав получившиеся результаты, можно сделать вывод, что наличие детей влияет на возврат кредита в срок. Клиенты без детей имеют меньше задолженностй по кредитам

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [89]:
debt = data.groupby('family_status')['debt'].sum()
total = data.groupby('family_status')['family_status'].count()
print(debt / total)


family_status
Не женат / не замужем    0.097842
в разводе                0.071247
вдовец / вдова           0.065539
гражданский брак         0.093142
женат / замужем          0.075575
dtype: float64


### Вывод

Проанализировав получившиеся результаты, можно сделать вывод, что лучше всего выплачивают кредит вдовцы и клиенты в разводе, чуть хуже - клиенты в браке. Хуже всего обстоят дела с клиентами, не бывавшими в браке и клиентами в гражданском браке.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [93]:
debt = data.groupby('grouping_total_income')['debt'].sum()
total = data.groupby('grouping_total_income')['grouping_total_income'].count()
print(debt / total)
data['total_income'].quantile([0.2, 0.4, 0.6, 0.8]).astype(int)

grouping_total_income
высокий доход          0.070184
доход выше среднего    0.083844
доход ниже среднего    0.083844
низкий доход           0.080782
средний доход          0.087376
dtype: float64


0.2     98526
0.4    132137
0.6    161370
0.8    214549
Name: total_income, dtype: int64

### Вывод

Проанализировав получившиеся результаты, можно сделать вывод, что клиенты с высоким доходом (более 200 тысяч) имеют меньше всего задержек по выплатам. Хуже всего выплачивают кредиты клиенты со средним доходом (от 130 до 160 тысяч).

- Как разные цели кредита влияют на его возврат в срок?

In [91]:
data_pivot = data.pivot_table(index=['category_purpose'], columns='debt', values='total_income', aggfunc='count')
data_pivot.set_axis(['without_debt','with_debt'],axis = 'columns',inplace = True)
data_pivot['part_debt'] = data_pivot['with_debt'] / (data_pivot['with_debt'] + data_pivot['without_debt'])
data_pivot

without_debt  with_debt  part_debt
category_purpose                                    
автомобиль                3861        397   0.093236
недвижимость              9926        777   0.072596
образование               3601        369   0.092947
свадьба                   2118        181   0.078730

### Вывод

Поанализировав получившиеся результаты, можно сделать вывод, что клиенты, которые берут кредит на недвижимость, имеют меньше всего задержек по выплатам. Клиенты, которые берут кредит на автомобиль или образование, хуже всех платят кредиты.

### Шаг 4. Общий вывод

Анализ предоставленнных данных показывает, что наличие детей, семейное положение и цель кредиты влияют на возврат кредита в срок.
Можно дать следующие рекомендации:
1. Лучшим клиентом является клиент без детей, вдовец/вдова, с доходом более 200 тысяч, который берет кредит на недвижимость.
2. Наиболее худшим вариантом является клиент с 4 детьми, никогда не был в браке, с доходом от 130 до 160 тысяч, который берет кредит на автомобиль.